init/deviceref.go

```diff
@@ -45,27 +49,27 @@ func (d *deviceRef) matchesBlkInfo(info *blkInfo) bool {
 	}
 }
 
+func calculateDevName(parent string, partition int) string {
+	name := parent
+	// some drivers use 'p' prefix for the partition number. TODO: find out where it is codified.
+	if strings.HasPrefix(parent, "/dev/nvme") || strings.HasPrefix(parent, "/dev/mmcblk") {
+		name += "p"
+	}
+	name += strconv.Itoa(partition + 1) // devname partitions start with "1"
+	return name
+}
+
 // checks if the reference is a gpt-specific and if yes then tries to resolve it to a device name
 func (d *deviceRef) resolveFromGptTable(devPath string, t []gptPart) *deviceRef {
-	if d.format != refGptType && d.format != refGptUUID && d.format != refGptLabel && d.format != refGptUUIDPartoff {
+	if d.format != refEspUUID && d.format != refGptUUID && d.format != refGptLabel && d.format != refGptUUIDPartoff {
 		return d
 	}
 
-	calculateDevName := func(parent string, partition int) string {
-		name := parent
-		// some drivers use 'p' prefix for the partition number. TODO: find out where it is codified.
-		if strings.HasPrefix(parent, "/dev/nvme") || strings.HasPrefix(parent, "/dev/mmcblk") {
-			name += "p"
-		}
-		name += strconv.Itoa(partition + 1) // devname partitions start with "1"
-		return name
-	}
-
 	for _, p := range t {
 		switch d.format {
-		case refGptType:
-			if bytes.Equal(d.data.(UUID), p.typeGUID) {
-				return &deviceRef{refName, calculateDevName(devPath, p.num)}
+		case refEspUUID:
+			if bytes.Equal(d.data.(UUID), p.uuid) {
+				return d.autodetectRoot(devPath, t)
```

Comment:
The `autodetectRoot() function does not accept a partition but instead gets whole table. For me it looks like the `autodetectRoot()` should be split into 2 parts

1. calculate `rootTypeGuid` - move outside of the loop
2. `if bytes.Equal(rootTypeGuid, p.typeGUID) {` - should be called inside the loop